polar-core/src/kb.rs

```diff
@@ -75,11 +81,247 @@ impl KnowledgeBase {
+                (
+                    Value::Variable(_),
+                    Some(Value::Pattern(template_spec)),
+                    Value::Variable(_),
+                    Some(rule_value),
+                )
+                | (Value::Variable(_), Some(Value::Pattern(template_spec)), rule_value, None) => {
+                    match template_spec {
+                        // Template specializer is an instance pattern
+                        Pattern::Instance(InstanceLiteral {
+                            tag,
+                            fields: template_fields,
+                        }) => match rule_value {
+                            Value::String(_) => tag == &sym!("String"),
+                            Value::Number(Numeric::Integer(_)) => tag == &sym!("Integer"),
+                            Value::Number(Numeric::Float(_)) => tag == &sym!("Float"),
+                            Value::Boolean(_) => tag == &sym!("Boolean"),
+                            Value::List(_) => tag == &sym!("List"),
+                            Value::Dictionary(rule_fields) => {
+                                tag == &sym!("Dictionary")
+                                    && self.param_fields_match(template_fields, rule_fields)
+                            }
+                            _ => {
+                                unreachable!("Value variant {} cannot be a specializer", rule_value)
+                            }
+                        },
+                        // Template specializer is a dictionary pattern
+                        Pattern::Dictionary(template_fields) => {
+                            if let Value::Dictionary(rule_fields) = rule_value {
+                                self.param_fields_match(template_fields, rule_fields)
+                            } else {
+                                false
+                            }
+                        }
+                    }
+                }
+
+                // If template has specializer but rule doesn't, then the rule does not match
+                (Value::Variable(_), Some(_), Value::Variable(_), None) => false,
+                // Template has no specializer, then the rule always matches
+                (Value::Variable(_), None, _, _) => true,
+                // Template has value specializer, rule has value specializer |
+                // template has value specializer, rule has value |
+                // template has value, rule has value
+                (
+                    Value::Variable(_),
+                    Some(template_value),
+                    Value::Variable(_),
+                    Some(rule_value),
+                )
+                | (Value::Variable(_), Some(template_value), rule_value, None)
+                | (template_value, None, rule_value, None) => match (template_value, rule_value) {
+                    (Value::List(template_list), Value::List(rule_list)) => template_list
```

Comment:
What's the logic here? Why are we not checking that the lists are equal?

Reply:
This is the `isa` logic that the list needs to contain at least all the elements of the other list in order to match, but could be a superset.